File: content/snippets/python/interview_lcci_08_07_permutation_i.py
from typing import List

def permutation(S: str) -> List[str]:
    result = []
    chars = sorted(S)
    used = [False] * len(chars)

    def backtrack(path):
        if len(path) == len(chars):
            result.append("".join(path))
            return
        for i, c in enumerate(chars):
            if not used[i]:
                used[i] = True
                path.append(c)
                backtrack(path)
                path.pop()
                used[i] = False

    backtrack([])
    return result

def main():
    for p in permutation("qwe"):
        print(p)

if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci_08_07_permutation_i.ts
function permutation(S: string): string[] {
    const chars = S.split("").sort();
    const result: string[] = [];
    const used = new Array(chars.length).fill(false);
    const path: string[] = [];

    function backtrack(): void {
        if (path.length === chars.length) {
            result.push(path.join(""));
            return;
        }
        for (let i = 0; i < chars.length; i++) {
            if (!used[i]) {
                used[i] = true;
                path.push(chars[i]);
                backtrack();
                path.pop();
                used[i] = false;
            }
        }
    }

    backtrack();
    return result;
}

for (const p of permutation("qwe")) {
    console.log(p);
}
